File: 3776.minimum-moves-to-balance-circular-array.java
#
# @lc app=leetcode id=3776 lang=java
#
# [3776] Minimum Moves to Balance Circular Array
#

# @lc code=start
class Solution {
    public long minMoves(int[] balance) {
        int n = balance.length;
        long sum = 0;
        int negIdx = -1;
        long need = 0;
        for (int i = 0; i < n; i++) {
            sum += balance[i];
            if (balance[i] < 0) {
                negIdx = i;
                need = -(long) balance[i];
            }
        }
        if (sum < 0) {
            return -1;
        }
        if (need == 0) {
            return 0;
        }
        long[] supply = new long[n];
        for (int i = 0; i < n; i++) {
            if (i == negIdx) continue;
            int dist1 = (i - negIdx + n) % n;
            int dist2 = (negIdx - i + n) % n;
            int dist = Math.min(dist1, dist2);
            supply[dist] += (long) balance[i];
        }
        long ans = 0;
        for (int d = 1; d < n; d++) {
            long take = Math.min(need, supply[d]);
            ans += take * d;
            need -= take;
            if (need <= 0) break;
        }
        return ans;
    }
};
# @lc code=end

File: 3776.minimum-moves-to-balance-circular-array.go
#
# @lc app=leetcode id=3776 lang=golang
#
# [3776] Minimum Moves to Balance Circular Array
#

# @lc code=start
func minMoves(balance []int) int64 {
	n := len(balance)
	if n == 0 {
		return 0
	}

	var total int64
	var deficit int64
	negPos := -1

	for i := 0; i < n; i++ {
		total += int64(balance[i])
		if balance[i] < 0 {
			negPos = i
			deficit = int64(-balance[i])
		}
	}

	if total < 0 {
		return -1
	}
	if deficit == 0 {
		return 0
	}

	supply := make([]int64, n)

	for i := 0; i < n; i++ {
		if i == negPos {
			continue
		}
		d := i - negPos
		if d < 0 {
			d = -d
		}
		md := d
		if n - d < md {
			md = n - d
		}
		supply[md] += int64(balance[i])
	}

	var ans int64
	rem := deficit

	for dist := 1; dist < n; dist++ {
		if rem <= 0 {
			break
		}
		take := supply[dist]
		if take > rem {
			take = rem
		}
		ans += take * int64(dist)
		rem -= take
	}

	if rem > 0 {
		return -1
	}

	return ans
}
# @lc code=end